### DISCUSSION TITLE
Add an alternative way to declare routes

### DISCUSSION THREAD

**Discussion by RafidMuhymin** [withastro/roadmap#68] [Proposal]
It would be a good addition to add an alternative way to declare routes using a `routes.js` or `routes.astro` file. Not only this will simplify routing but it can unlock a lot of advanced use cases which is currently not possible (or hard to achieve) using *Filesystem-based Routing*. The *imaginary* file may have the following features:

1. The frontmatter will be available like all the other Astro components.
2. You'll be able to use only <Route> components inside a fragment (not required in Astro)
3. There'll be a `<Route />` component which will take the path as a prop and be able to receive content via slot
4. You'll be able to map over a frontmatter array and create a `<Route />` component for each of them (Dynamic Routing)
5. (Extra) It isn't required to import the `<Route />` component to use them (The compiler or Vite will do this by default)

**Comment by f3oall** [comment]
This is a good idea to have alternatives.

On one project we have a pretty messed up URL structure. It would be cool to have the ability to structure src code differently. But Astro should be able to have custom routing customization in this case.

Also in the same project, all our URLs have a .htm extension. We can't implement redirects because of SEO reasons. So it would be nice to choose result file extensions as well.

**Comment by dobby** [comment]
Seo is the main issue here, complex seo requirements are more then a fixed folder structure. Example: a product that has the category structure in its url path /cat1/cat2/cat3/productName-code. The depth of the cat structure is not known in advance and the customer can freely decide in the CMS what structure he wants in the url so url structure is not known at build time. So yes dynamic programmatic routing is a must.
